sphinx_needs/api/need.py

```diff
@@ -770,48 +689,56 @@ def make_hashed_id(
     return f"{type_prefix}{cal_hashed_id[:id_length]}"
 
 
-def _fix_list_dyn_func(list: list[str]) -> list[str]:
-    """
-    This searches a list for dynamic function fragments, which may have been cut by generic searches for ",|;".
-
-    Example:
-    `link_a, [[copy('links', need_id)]]` this will be splitted in list of 3 parts:
-
-    #. link_a
-    #. [[copy('links'
-    #. need_id)]]
+def _split_list_with_dyn_funcs(
+    text: None | str | list[str], location: tuple[str, int | None] | None
+) -> list[str]:
+    """Split a ``;|,`` delimited string that may contain ``[[...]]`` dynamic functions.
 
-    This function fixes the above list to the following:
+    Remove any empty strings from the result.
 
-    #. link_a
-    #. [[copy('links', need_id)]]
-
-    :param list: list which may contain splitted function calls
-    :return: list of fixed elements
+    If the input is a list of strings, return the list unchanged,
+    or if the input is None, return an empty list.
     """
-    open_func_string = False
-    new_list = []
-    for element in list:
-        # If dyn_func got not cut, just add it
-        if "[[" in element and "]]" in element:
-            new_list.append(element)
-        # Other check if this is the starting element of dyn function
-        elif "[[" in element:
-            open_func_string = True
-            new_link = [element]
-        # Check if this is the ending element if dyn function
-        elif "]]" in element:
-            new_link.append(element)
-            open_func_string = False
-            element = ",".join(new_link)
-            new_list.append(element)
-        # Check it is a "middle" part of the dyn function
-        elif open_func_string:
-            new_link.append(element)
-        # Looks like it isn't a cut dyn_func, just add.
+    if text is None:
+        return []
+
+    if not isinstance(text, str):
+        assert isinstance(text, list) and all(
+            isinstance(x, str) for x in text
+        ), "text must be a string or a list of strings"
+        return text
+
+    result: list[str] = []
+    _current_element = ""
+    while text:
+        if text.startswith("[["):
+            _current_element += text[:2]
+            text = text[2:]
+            while text and not text.startswith("]]"):
+                _current_element += text[0]
+                text = text[1:]
+            if _current_element.endswith("]"):
+                _current_element += "]"
+            else:
+                _current_element += "]]"
+            if not text.startswith("]]"):
+                log_warning(
```

Comment:
Feel closing the brackets on SN side is dangerous as it might not be what the user wants/expects. I'd either
- give up and return an empty string with a warning or
- return to the outer loop and go on eating tokens starting from `[[` (means it's not a dynamic function if no closing token is found) but still emit a warning.

Reply:
Fine for me. `[[` is just always a dynamic function and (without backtracking) we just go to the end. If this is not intended by the user, it will most certainly fail as the function does not exist.